src/backend/vulkan/RenderPassCache.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BACKEND_VULKAN_RENDERPASSCACHE_H_
+#define BACKEND_VULKAN_RENDERPASSCACHE_H_
+
+#include "common/vulkan_platform.h"
+
+#include "common/Constants.h"
+#include "nxt/nxtcpp.h"
+
+#include <array>
+#include <bitset>
+#include <unordered_map>
+
+namespace backend { namespace vulkan {
+
+    class Device;
+
+    // This is a key to query the RenderPassCache, it can be sparse meaning that only the
+    // information for bits set in colorMask or hasDepthStencil need to be provided and the rest can
+    // be uninintialized.
+    struct RenderPassCacheQuery {
+        // Use these helpers to build the query, they make sure all relevant data is initialized and
+        // masks set.
+        void SetColor(uint32_t index, nxt::TextureFormat format, nxt::LoadOp loadOp);
+        void SetDepthStencil(nxt::TextureFormat format,
+                             nxt::LoadOp depthLoadOp,
+                             nxt::LoadOp stencilLoadOp);
+
+        std::bitset<kMaxColorAttachments> colorMask;
+        std::array<nxt::TextureFormat, kMaxColorAttachments> colorFormats;
+        std::array<nxt::LoadOp, kMaxColorAttachments> colorLoadOp;
+
+        bool hasDepthStencil = false;
+        nxt::TextureFormat depthStencilFormat;
+        nxt::LoadOp depthLoadOp;
+        nxt::LoadOp stencilLoadOp;
+    };
+
+    // Caches VkRenderPasses so that we don't create duplicate ones for every RenderPipeline or
+    // render pass.
+    // TODO(cwallez@chromium.org): Make it an LRU cache somehow?
```

Comment:
Out of curiosity, does this mean it grows without bound, currently?

Reply:
Yes but given the small amount of real-world permutations that shouldn't cause problems before V1. For example in Skia's case the cache could grow to a whopping 10 elements!